File: sw-expert-academy/D3/햄버거 다이어트.java
import java.io.IOException;
import java.util.Arrays;
import java.util.Scanner;
 
public class Solution {
 
    static int answer;
    static int T, N, L;
    static int Ti, Ki; // 맛 점수, 칼로리
    static int[] hamburger; // 칼로리(인덱스), 맛 점수(값)
 
    public static void main(String[] args) throws IOException {
 
        Scanner sc = new Scanner(System.in);
        T = sc.nextInt();
 
        for (int tc = 1; tc <= T; tc++) {
            // initialize
            N = sc.nextInt();
            L = sc.nextInt();
            hamburger = new int[L + 1];
 
            // solution
            for (int i = 0; i < N; i++) {
                Ti = sc.nextInt();
                Ki = sc.nextInt();
                for (int j = L; j >= Ki; j--) 
                    hamburger[j] = Math.max(hamburger[j], hamburger[j - Ki] + Ti);
            }
            answer = Arrays.stream(hamburger).max().getAsInt(); 
             
            // result
            System.out.printf("#%d %d\n", tc, answer);
        }
    }
 
}

File: sw-expert-academy/D3/햄버거 다이어트.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {

	int T;
	cin >> T;

	for (int tc = 1; tc <= T; tc++) {
		int answer = 0;
		int N, L, flavor, calorie;
		cin >> N >> L;

		vector<int> hamburger(L + 1, -1); // 칼로리당 맛점수를 표시
		hamburger[0] = 0;

		for (int i = 0; i < N; i++) {
			cin >> flavor >> calorie;
			for (int j = L; j >= calorie; j--)
				if (hamburger[j - calorie] >= 0)
					hamburger[j] = max(hamburger[j], flavor + hamburger[j - calorie]);
		}

		// 가장 맛 점수가 높은 햄버거 구하기
		answer = *max_element(hamburger.begin(), hamburger.end());
		cout << '#' << tc << ' ' << answer << '\n';
	}

	return 0;
}